File: go/s0101/main.go
package main

import (
	"context"
	"github.com/PuerkitoBio/goquery"
	"github.com/joho/godotenv"
	"github.com/openai/openai-go"
	"github.com/openai/openai-go/option"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
)

func main() {
	ctx := context.Background()
	err := godotenv.Load("../../.env")
	if err != nil {
		log.Fatalf("could not load env variables %v", err)
	}

	openaiClient := openai.NewClient(option.WithAPIKey(os.Getenv("OPENAI_API_KEY")))
	client := http.Client{}

	host := os.Getenv("HOST")
	resp, err := client.Get(host)

	if err != nil {
		log.Fatalln("could not fetch resources from hot", err)
	}

	if resp.StatusCode == http.StatusOK {
		question := extractQuestion(resp.Body)
		answer := callModelForAnswer(ctx, openaiClient, question)
		values := url.Values{}
		values.Add("username", os.Getenv("AGENT_USER"))
		values.Add("password", os.Getenv("AGENT_PASSWORD"))
		values.Add("answer", answer)
		finalResp, err := client.PostForm(host, values)
		if err != nil {
			log.Fatalln("something went wrong while logging to the system")
		}
		if finalResp.StatusCode == http.StatusOK {
			log.Print("logging with success")
			result, err := io.ReadAll(finalResp.Body)
			if err != nil {
				log.Fatalf("could not read body %v", err)
			}
			log.Printf(string(result))
		}
	}
}

func extractQuestion(closer io.ReadCloser) string {
	doc, err := goquery.NewDocumentFromReader(closer)
	if err != nil {
		log.Fatalf("could not load html %v", err)
	}

	questionParagraph := doc.Find("p#human-question")

	html, err := questionParagraph.Html()
	if err != nil {
		log.Fatalf("error extracting paragraph from HTML: %v", err)
	}

	parts := strings.Split(html, "<br/>")
	if len(parts) < 2 {
		log.Fatalf("Unexpected format in human-question paragraph")
	}
	question := strings.TrimSpace(parts[1])
	log.Printf("extracted question %s", question)
	return question
}

func prepareSystemMessage() openai.ChatCompletionMessageParam {
	return openai.ChatCompletionMessageParam{
		Role:    openai.F(openai.ChatCompletionMessageParamRole("system")),
		Content: openai.F[interface{}]("You will receive the question. You need to answer as short as possible, the best answer is 1 word if possible."),
	}
}

func prepareUserMessage(question string) openai.ChatCompletionMessageParam {
	return openai.ChatCompletionMessageParam{
		Role:    openai.F(openai.ChatCompletionMessageParamRole("user")),
		Content: openai.F[interface{}](question),
	}
}

func callModelForAnswer(ctx context.Context, client *openai.Client, question string) string {
	chatCompletion, err := client.Chat.Completions.New(ctx, openai.ChatCompletionNewParams{
		Messages: openai.F([]openai.ChatCompletionMessageParamUnion{prepareSystemMessage(), prepareUserMessage(question)}),
		Model:    openai.F(openai.ChatModelGPT4oMini),
	})

	if err != nil {
		log.Fatalln("error while calling openai", err)
	}
	answer := chatCompletion.Choices[0].Message.Content
	log.Printf("answer -> %s", answer)
	return answer
}

File: python/s0101/main.py
import os
import sys

import requests
from openai import OpenAI
from dotenv import load_dotenv
from pathlib import Path
import logging
from bs4 import BeautifulSoup as BS

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    handlers=[
        logging.StreamHandler(sys.stdout)
    ]
)


def prepare_system_message():
    return {
        "role": "system",
        "content": "You will receive the question. You need to answer as short as possible, the best answer is 1 word if possible."
    }

def prepare_user_message(question: str):
    return {
        "role": "user",
        "content": question
    }

def extract_question(html: str):
    soup = BS(html, "html.parser")
    p_tag = soup.find('p', id="human-question")

    if p_tag:
        question = p_tag.get_text(separator=' ').replace('Question:', '').strip()
        logging.info(f"question is: {question}")
        return question
    return None


def call_model_for_answer(question: str):
    client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY"))

    chat_completion = client.chat.completions.create(
        messages=[prepare_system_message(), prepare_user_message(question)],
        model="gpt-4o-mini",
    )
    answer = chat_completion.choices[0].message.content
    logging.info(f"answer is: {answer}")
    return answer

def main():
    dotenv_path = Path('../../.env')
    load_dotenv(dotenv_path=dotenv_path)
    host = os.environ.get("HOST")
    resp = requests.get(host)

    if resp.status_code != 200:
        logging.error(f"something went wrong while calling host {host}")
        sys.exit()

    question = extract_question(resp.content)
    if question is None:
        logging.error("failed to extract question")
        sys.exit()

    answer = call_model_for_answer(question)

    payload = {'username':os.environ.get("AGENT_USER"),'password':os.environ.get("AGENT_USER"), 'answer': answer}
    response = requests.post(host, files=payload)
    if response.status_code != 200:
        logging.error("something went wrong while logging to the system", response.content)
        sys.exit()

    if response.status_code == 200:
        logging.info("logging with success")

if __name__ == '__main__':
    main()